MobiusCore/Source/EffectHandlers/EffectExecutor.swift

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2019 Spotify AB.
+//
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import Foundation
+
+class EffectExecutor<Effect, Event>: Connectable {
+    private let handleEffect: (Effect, Response<Event>) -> Disposable
+    private var output: Consumer<Event>?
+
+    private let lock = Lock()
+
+    // Keep track of each received effect's state.
+    // When an effect has completed, it should be removed from this dictionary.
+    // When disposing this effect handler, all entries must be removed.
+    private var handlingEffects: [Int64: EffectHandlingState<Event>] = [:]
+    private var nextID = Int64(0)
+
+    init(handleInput: @escaping (Effect, Response<Event>) -> Disposable) {
+        self.handleEffect = handleInput
+    }
+
+    func connect(_ consumer: @escaping Consumer<Event>) -> Connection<Effect> {
+        output = consumer
+        return Connection(
+            acceptClosure: handle,
+            disposeClosure: dispose
+        )
+    }
+
+    func handle(_ effect: Effect) {
+        let id: Int64 = lock.synchronized {
+            nextID += 1
+            return nextID
+        }
+
+        let response = Response(
+            // Any events produced as a result of handling the effect will be sent to this class's `output` consumer,
+            // unless it has already been disposed.
+            onSend: { [weak self] event in self?.output?(event) },
+            // Once an effect has been handled, remove the reference to its response and disposable.
+            onEnd: { [weak self] in self?.delete(id: id) }
+        )
+
+        let disposable = handleEffect(effect, response)
+
+        // If the effect has not yet been handled, store a reference to its response and disposable.
```

Comment:
This is actually racy - it works for the synchronous case, but may break for 'fast async' cases. That is, `Response.end` may be invoked between the check of 'ended' and the call to 'store'. 

I think it's safe to write this instead:

```
store(id: id, response: response, disposable: disposable)
if response.ended {
    delete(id: id)
}
```

Since duplicate deletions should be fine.